src/unity.c

```diff
@@ -235,97 +235,121 @@ void UnityPrintMask(const UNITY_UINT mask, const UNITY_UINT number)
+ *  else if (number < 0.1 && number > -0.1)              snprintf(buf, sizeof buf, "%.5e", number);
+ *  else                                                 snprintf(buf, sizeof buf, "%.6f", number);
+ *  UnityPrint(buf);
+ */
 void UnityPrintFloat(const UNITY_DOUBLE input_number)
 {
     UNITY_DOUBLE number = input_number;
 
     /* print minus sign (including for negative zero) */
-    if (number < 0.0f || (number == 0.0f && 1.0f / number < 0.0f))
+    if (number < 0 || (number <= 0 && 1.0f / number < 0))
     {
         UNITY_OUTPUT_CHAR('-');
         number = -number;
     }
 
-    /* handle zero, NaN, and +/- infinity */
-    if (number == 0.0f) UnityPrint("0");
-    else if (isnan(number)) UnityPrint("nan");
-    else if (isinf(number)) UnityPrint("inf");
-    else
+    if (isnan(number)) UnityPrint(UnityStrNaN);
+    else if (isinf(number)) UnityPrintLen(UnityStrInf, 3);
+    else if (number < 0.1 && number > 0)
     {
+        /* Prints only 6 significant digits. */
         int exponent = 0;
-        int decimals, digits;
         UNITY_INT32 n;
-        char buf[16];
 
-        /* scale up or down by powers of 10 */
+        /* scale up by powers of 10 */
         while (number < 100000.0f / 1e6f)  { number *= 1e6f; exponent -= 6; }
         while (number < 100000.0f)         { number *= 10.0f; exponent--; }
-        while (number > 1000000.0f * 1e6f) { number /= 1e6f; exponent += 6; }
-        while (number > 1000000.0f)        { number /= 10.0f; exponent++; }
 
         /* round to nearest integer */
-        n = ((UNITY_INT32)(number + number) + 1) / 2;
+        n = (UNITY_INT32)(number + 0.5f);
         if (n > 999999)
         {
             n = 100000;
             exponent++;
         }
 
-        /* determine where to place decimal point */
-        decimals = (exponent <= 0 && exponent >= -9) ? -exponent : 5;
-        exponent += decimals;
+        exponent += 5; /* Correct for decimal place */
+        exponent = -exponent; /* Exponent is always negative in this scope */
 
-        /* truncate trailing zeroes after decimal point */
-        while (decimals > 0 && n % 10 == 0)
-        {
-            n /= 10;
-            decimals--;
-        }
+        UNITY_OUTPUT_CHAR('0' + n / 100000);
+        UnityPrintDecimalAndNumberWithLeadingZeros(n % 100000, 100000 / 10);
+        UNITY_OUTPUT_CHAR('e');
+        UNITY_OUTPUT_CHAR('-');
+        if (exponent < 10) UNITY_OUTPUT_CHAR('0');
+        UnityPrintNumber(exponent);
+    }
+    else if (number < 4294967295.9999995) /* Rounded result fits in 32 bits, "%.6f" format */
+    {
+        const UNITY_INT32 divisor = 1000000/10;
+        UNITY_UINT32 integer_part = (UNITY_UINT32)number;
+        UNITY_INT32 fraction_part = (UNITY_INT32)((number - (UNITY_DOUBLE)integer_part)*1000000.0 + 0.5);
+        /* Double precision calculation gives best performance for six rounded decimal places */
```

Comment:
Can we/do we want to rely on having double-precision support?

Reply:
Definitely no. A lot of the supported targets don't have double precision... and we're unlikely to ever print that many significant digits anyway. ;)

Reply:
It's another trade-off, as folks mentioned above. You need more bits of precision to accurately print more digits. With a 23 bit mantissa the calculation cannot even round the 6th digit correctly a lot of the time. Surprising, huh.
You can eliminate the use of doubles by excluding the code. `UNITY_EXCLUDE_FLOAT_PRINT` handles that option for tight embedded systems.